Adds translators between modeling libraries and quadratic program

t-imamichi opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add it in the CHANGELOG file under Unreleased section.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary

Revised version of #116.

It introduces a model translators to convert a modeling libraries (e.g., docplex, gurobipy, etc.) to `QuadraticProgram` and vice versa.

It adds `qiskit_optimization.translators` directory and the following functions are implemented.
- `from_docplex_mp`
- `to_docplex_mp`
- `from_gurobipy`
- `to_gurobipy`

Because I plan to implement `docplex.cp` version of translators in the future, I use the name `docplex_mp` instead of just `docplex`.

Fixes #80

TODO
- [x] class design
- [x] naming of methods
- [x] unit tests
- [x] Gurobi translator
- [x] update unit tests that use `QP.from_docplex` or `QP.to_docplex`.
- [x] update tutorials that use `QP.from_docplex` or `QP.to_docplex`. 

### Details and comments

Deprecations
- `QuadraticProgram.from_docplex`: will be replaced with `from_docplex_mp`.
- `QuadraticProgram.to_docplex`: will be replaced with `to_docplex_mp`

Removal
- `QuadraticProgram.{pprint_as_string,prettyprint}`: they were deprecated in aqua as mentioned https://github.com/Qiskit/qiskit-optimization/pull/122#discussion_r650233810

t-imamichi commented on qiskit_optimization/translators/model_translator.py:
```diff
@@ -0,0 +1,64 @@
+        Returns:
+            Returns True if the model is compatible, False otherwise.
+        """
+        pass
+
+    @abstractmethod
+    def qp_to_model(self, quadratic_program: "QuadraticProgram") -> Any:
+        """Returns an optimization model corresponding to a quadratic program.
+
+        Args:
+            quadratic_program: The quadratic program to be translated
+
+        Returns:
+            The optimization model corresponding to a quadratic program.
+        """
+        pass
+
+    @abstractmethod
+    def model_to_qp(self, model: Any) -> "QuadraticProgram":
```
I don't think `Translator.{load,save}` make sense because the direction is not clear. Do you have any intuitive naming idea?

woodsp-ibm replied:
Since these methods are in model class having the name model in them may not be necessary - how about simply `from_qp` and `to_qp` where the from and to are implicitly model since they are methods of the class.

t-imamichi replied:
`from_qp` and `to_qp` sound good too.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -869,6 +871,40 @@ def maximize(
             self, constant, linear, quadratic, QuadraticObjective.Sense.MAXIMIZE
         )
 
+    def load(self, model: Any) -> "QuadraticProgram":
```
I'm wondering whether we need a `load` method for `QuadraticProgram` because `utils.load_model` might be sufficient. What do you think?

adekusar-drl replied:
I don't think we should expose anything from `utils.*` to end users. Less classes/names/etc to remember - easier to understand. So, I think a static method should be a good choice. It is like `QuantumCircuit.from_qasm_file()`.

t-imamichi replied:
I see. Then, it would be better to rename `utils.load_model` with `utils._load_model`. I agree to make `QuadraticProgram.load` static.

adekusar-drl replied:
Good point to make it private!

adekusar-drl commented on qiskit_optimization/translators/utils.py:
```diff
@@ -0,0 +1,48 @@
+from qiskit_optimization.exceptions import QiskitOptimizationError
+from .docplex import DocplexMpTranslator
+
+if TYPE_CHECKING:
+    # pylint: disable=cyclic-import
+    from qiskit_optimization.problems.quadratic_program import QuadraticProgram
+
+translators = [DocplexMpTranslator()]
```
Well, I mean to have something like this:
```python
translators = {Model: DocplexMpTranslator()}

# create a DOcplex model
some_model = Model()

# load the model
qp = translators[type(some_model)].model_to_qp(model)
```
This would just remove the loop.

adekusar-drl replied:
Oh, I see, then maybe it is easier to keep the list approach!

woodsp-ibm commented on qiskit_optimization/translators/model_translator.py:
```diff
@@ -0,0 +1,64 @@
+    """
+
+    @abstractmethod
+    def is_compatible(self, model: Any) -> bool:
```
In the docplex case we have that installed always right as a dependent. If I pass a translator in a list, I am looking at code below that goes over the list checking is_compatible and is so translates it, if the necessary optimization model code is part of an optional install, what do we expect to happen. Should the translator fail to load and throw an exception as the dependent is not installed - or do we expect the user to pass in the list as some sort of preferred order of translation whereby we pick the one that works i.e the translator will still load but say its not compatible, not because its cannot truly do it, but because the optional dependence it needs is not installed and thus is not presently compatible until such time, if ever, that were changed.

t-imamichi replied:
If users try to translate an optimization model of a dependent library, the model is made by the library. So the library is supposed to be installed in their environment and the translator can catch the model. However, it might be convenient to have a method `is_installed` to check whether the library is installed or not.

stefan-woerner commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -869,6 +871,41 @@ def maximize(
+        """
+        return _load_model(model)
+
+    def save(self, translator: Optional[ModelTranslator] = None) -> Any:
```
should we rename this to `export` and then maybe just incorporate the `export_to_lp_string` as well via `translator='lp'` or something like that?

t-imamichi replied:
Thank you for your idea. `export_to_lp_string` came from docplex method. If we make docplex translator, it would be simpler to make `lp` as a translator.

I'm wondering the pair of names "load/save", "?/export", or "from_qp/to_qp". Word "import" and "from" are not allowed because they are reserved by Python syntax.

t-imamichi replied:
I notice that we can deprecate other methods as well https://github.com/Qiskit/qiskit-optimization/pull/122#issuecomment-849341955

t-imamichi replied:
I can add a `LPStringTranslator` that supports only `from_qp(qp) -> str` and does not support `to_qp(model) -> QP`. What do you think?

t-imamichi replied:
Do you prefer naming `export` and `load`?

t-imamichi replied:
I implemented your idea. Could you verify it?
https://github.com/Qiskit/qiskit-optimization/pull/122#discussion_r640105391

stefan-woerner commented on qiskit_optimization/translators/__init__.py:
```diff
@@ -0,0 +1,42 @@
+   :toctree: ../stubs/
+   :nosignatures:
+
+   DocplexTranslator
```
update name

stefan-woerner commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1034,81 +1075,10 @@ def to_docplex(self) -> Model:
-                raise QiskitOptimizationError("Unsupported constraint sense: {}".format(sense))
-
-        return mdl
+        # warnings.warn("The to_docplex method is deprecated and will be "
```
update string to new function name and raise deprecation

t-imamichi commented:
I notice that we can simplify `QP.read_from_lp_file` and `QP.write_to_lp_file` by making a translator. I will work on it.

t-imamichi commented:
I introduced `LPFileTranslator` to replace `read_from_lp_file` and `write_to_lp_file`. What do you think?

woodsp-ibm commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1286,89 +939,27 @@ def to_docplex(self) -> Model:
+        )
+        return self.export()
 
     def export_as_lp_string(self) -> str:
```
I am wondering why not keep these methods and put what is in the string support part of export above here and let export just take a translator. The string values there will not autocomplete with an IDE so its a bit more discoverable this way. It would still be done by a translator; but I can understand wanting the symmetry of a  single load and single save like methods.

t-imamichi replied:
Yes. We should fix it, but I want leave it in this PR. How about discussing it at #77?

t-imamichi replied:
I tried @stefan-woerner's idea to allow `export('lp')`. https://github.com/Qiskit/qiskit-optimization/pull/122#discussion_r640105391
I see your point. It is hard to discover the method to convert qp to a string with an IDE.
I'm OK to revert this change, i.e., `export` allows only translators as option.

t-imamichi replied:
I reverted `qp.export_as_lp_string` and `qp.export`. If `qp.export` takes only translators, the impl gets simple. It is also easy for an IDE to discover `export_as_lp_string` too.

t-imamichi commented on qiskit_optimization/translators/__init__.py:
```diff
@@ -0,0 +1,46 @@
+   :toctree: ../stubs/
+   :nosignatures:
+
+   ModelTranslator
```
Thanks. You are right. I think of converting docplex or gurobi models to qp, so I named it `ModelTranslator`; but, I added a converter of a LP file to qp. Your idea makes sense.

t-imamichi commented on qiskit_optimization/translators/utils.py:
```diff
@@ -0,0 +1,51 @@
+translators = [DocplexMpTranslator(), GurobiTranslator(), LPFileTranslator()]
+
+
+def _load_model(model: Any) -> "QuadraticProgram":
```
@adekusar-drl suggested to hide `utils.*` from users. We thought `QP.load` is enough. https://github.com/Qiskit/qiskit-optimization/pull/122#discussion_r630091813

I agree to make `translators` private, `is_compatible` a class method. We can make `is_installed` a class method too.

t-imamichi replied:
I tried to make `_translators = [DocplexMpTranslator, ..., ]`, but mypy fails as follows. We need perhaps a trick of mypy. I avoid it by instantiate objects.
```
mypy qiskit_optimization test tools
qiskit_optimization/translators/utils.py:46: error: "ABCMeta" has no attribute "is_installed"
qiskit_optimization/translators/utils.py:46: error: "ABCMeta" has no attribute "is_compatible"
```

woodsp-ibm replied:
Maybe it needs to be typehinted as list of ModelTranslators

t-imamichi replied:
Thanks. I tried `_translator_types: List[Type[QuadraticProgramTranslator]]` and it works.

t-imamichi replied:
I'm thinking of moving the code of `_load_model` to `QP.load` because it is not so long. But, importing all translators in QP might be complicated. So, I may leave `_translator_types` in utils.py as a private list.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -108,7 +79,7 @@ def __init__(self, name: str = "") -> None:
         self._objective = QuadraticObjective(self)
 
     def __repr__(self) -> str:
-        return self.to_docplex().export_as_lp_string()
+        return self.export("lp")
```
Yes. We should fix it, but I want leave it in this PR. How about discussing it at #77?

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -886,144 +857,52 @@ def maximize(
-        - binary / integer / continuous variables
+    @staticmethod
+    def load(source: Any) -> "QuadraticProgram":
+        """Loads this quadratic program from an external source.
```
Yes. I will clarify it.

t-imamichi commented on qiskit_optimization/translators/docplex_mp.py:
```diff
@@ -0,0 +1,309 @@
+        return True
+
+    def is_compatible(self, model: Any) -> bool:
+        """Checks whether a given model can be translated with this translator.
```
Thanks. I use your message in the docstring.

t-imamichi commented on qiskit_optimization/translators/__init__.py:
```diff
@@ -0,0 +1,46 @@
+
+.. currentmodule:: qiskit_optimization.translators
+
+Translators between an optimization model and a quadratic program
```
Thanks. I revised the docstring.

t-imamichi commented on qiskit_optimization/translators/lp_file.py:
```diff
@@ -0,0 +1,110 @@
+            QiskitOptimizationError: if non-supported elements (should never happen).
+        """
+        mdl = DocplexMpTranslator().from_qp(quadratic_program)
+        mdl.export_as_lp(self._filename)
```
I made the constructor option `output_filename` optional and check whether it is `None` or not. It is safer than the previous one.

t-imamichi replied:
I reverted the LPFileTranslator because we should have more intuitive design dedicated for file I/O.

t-imamichi commented on qiskit_optimization/translators/lp_file.py:
```diff
@@ -0,0 +1,110 @@
+class LPFileTranslator(ModelTranslator):
+    """Translator between a LP file and a quadratic program"""
+
+    def __init__(self, output_filename: str = ""):
```
I made `output_filename` optional in the latest commit to check it `None` or not at `from_qp`.

t-imamichi commented:
Thank you for your feedback, @woodsp-ibm. Here is the summary of questions left. I welcome opinions of other people.

- Naming of methods: `QuadraticProgramTranslator` (QPT), `QPT.{from_qp, to_qp}`, `QuadraticProgram.{load,export}`.
- Shall we Include `LPFileTranslator` as part of QPT?
- Shall we allow `QP.export('lp')` to behave same as `QP.export_as_lp_string()`?
- Shall we make `qiskit_optimizations.translators.utils._load_qp_from_source` public or private?
- Need to revise `__repr__` (mentioned in #77) maybe in a separate PR.

adekusar-drl commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -886,144 +858,38 @@ def maximize(
-        # Retrieve the objective
-        objective = model.getObjective()
-        has_quadratic_objective = False
+    def export(self, translator: Optional[QuadraticProgramTranslator] = None) -> Any:
```
Should we have `load/save`, `read/write`? A combination of `export/load` does not look naturally.

t-imamichi replied:
This method basically returns a model of gurobi or docplex. Does `save` or `write` make sense?

t-imamichi replied:
I removed `export` and reverted `to_docpelx` and `to_gurobipy`. I think it is easy for IDE to discover.

adekusar-drl replied:
Does it mean if we want to have another translator, we add a new method like `to_zzz` to QP?
Let's discuss namings separately as agreed.

t-imamichi replied:
For example, pandas use such a design. https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_dict.html
Sure. I want to discuss the design and naming.

adekusar-drl reviewed (commented):
Looks good, I think namings could be better.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -886,144 +858,38 @@ def maximize(
-            linear = {}
-            quadratic = {}
+        if translator is None:
+            translator = DocplexMpTranslator()
```
I plan to implement `DocplexCpTranslator` for `docplex.cp`. I would like to keep `DocplexMp` to distinguish `docplex.mp` and `docplex.cp`.

adekusar-drl replied:
> 
> I plan to implement `DocplexCpTranslator` for `docplex.cp`. I would like to keep `DocplexMp` to distinguish `docplex.mp` and `docplex.cp`.

OK, then it makes sense, thanks.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1286,89 +926,21 @@ def to_docplex(self) -> Model:
 
         Returns:
             A string representing the quadratic program.
         """
-        return self.to_docplex().export_as_lp_string()
+        return self.export().export_as_lp_string()
```
You are right. Although we currently rely on docplex's `export_as_lp_string` for text representation, we need our own text representation (`__str__` and `__repr__`) as you mentioned #77. @woodsp-ibm also mentioned https://github.com/Qiskit/qiskit-optimization/pull/122#discussion_r646751522.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1377,12 +949,12 @@ def pprint_as_string(self) -> str:
             DeprecationWarning,
         )
-        return self.to_docplex().pprint_as_string()
+        return self.export().pprint_as_string()
```
I reverted `to_docplex`.

t-imamichi commented:
I tried to add `LPFileTranslator` as an example of translators. But, I found out the design is not so intuitive. We should discuss I/O framework in the future as a separate PR.

The current version includes the following changes:
- QuadraticProgramTranslator for docplex.mp and gurobipy
- QuadraticProgram.load to generatea a qp from either docplex.mp or gurobipy.

t-imamichi commented:
Now I simplified translators by making the following functions. I removed `QuadraticProgramTranslator`.
- `from_docplex_mp`
- `to_docplex_mp`
- `from_gurobipy`
- `to_gurobipy`

a-matsuo commented:
LGTM. I like making translators as functions. It's simpler and more intuitive (Users should know what library they are using).

woodsp-ibm commented on qiskit_optimization/translators/docplex_mp.py:
```diff
@@ -0,0 +1,286 @@
+        elif x.vartype == Variable.Type.INTEGER:
+            var[idx] = mdl.integer_var(lb=x.lowerbound, ub=x.upperbound, name=x.name)
+        else:
+            # should never happen
+            raise QiskitOptimizationError("Unsupported variable type: {}".format(x.vartype))
```
I wonder whether it would be better to just throw a RuntimeError given this "should never happen". If it does its some sort of coding error or whatever - in some internal error that the user most likely cannot do anything about to overcome this.  So perhaps the message could be "Internal error: unsupported variable type:. {}"

Same for contrainst below.

t-imamichi replied:
You are right. This is internal error that should not happen. So, It makes sense to change it `RuntimeError("Internal error: ...")` for `to_docplex_mp`.

woodsp-ibm commented on qiskit_optimization/translators/docplex_mp.py:
```diff
@@ -0,0 +1,286 @@
+        elif isinstance(x.vartype, IntegerVarType):
+            x_new = quadratic_program.integer_var(x.lb, x.ub, x.name)
+        else:
+            raise QiskitOptimizationError(
```
This on the other hand seems fine - we can potentially expect this to occur since the docplex model has a range of capability that is not supported by QP. In this case the user should adjust their model to remove such things. 

Question though - is there ever a case where we would want to say log a warning and continue, just taking from the model the bits we do understand and ignoring the others. Maybe its meaningless - if it made sense then perhaps an `ignore_unsupported=False` param on the method, allowing the behavior to be changed between failing hard like it does now or continuing albeit that aspects get dropped.

t-imamichi replied:
If we ignore unsupported types of elements (variables, constraints, etc.) in `from_docplex_mp`, the resulting QP model will be different and the solution will be also different from what users expect. So, I would like to keep the current behavior to raise an error with unsupported type for `from_docplex_mp`.

woodsp-ibm commented:
Will you also update the tutorials? They show docplex model conversion etc. and we should show this the new way now I think, as otherwise they will have deprecation warnings in them that we try and avoid.

t-imamichi commented:
Thank you for your review. I fix tutorials and unit tests that uses `QP.{from,to}_docplex`.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1286,89 +893,26 @@ def to_docplex(self) -> Model:
+
+        return to_docplex_mp(self).export_as_lp_string()
 
     def pprint_as_string(self) -> str:
```
Thanks. It is a good timing to remove them. I originally plan to do it #153, but, it is good to do it as part of this PR.

t-imamichi commented on qiskit_optimization/problems/quadratic_program.py:
```diff
@@ -1039,243 +871,18 @@ def from_docplex(self, model: Model) -> None:
             QiskitOptimizationError: if the model contains unsupported elements.
         """
+        warnings.warn(
+            "The from_docplex method is deprecated and will be "
```
Thanks. I updated it to mention "in a future release" -> "no sooner than 3 months after the release".

woodsp-ibm replied:
Can we include the version in there, from when this is deprecated, so people can better figure out when etc - and also we can more easily keep track too as to when to delete the code e.g. to the first line below, one example, I added `as of version 0.2.0` 

So from
```
"The from_docplex method is deprecated and will be "
...
```
to
```
"The from_docplex method is deprecated as of version 0.2.0 and will be "
"removed no sooner than 3 months after the release. Instead use the "
"qiskit_optimization.translators.from_docplex_mp function"
```

t-imamichi replied:
Sure. I fixed it.

t-imamichi commented:
I have updated unit tests and tutorials so that everything uses the new `from_docplex_mp` and `to_docplex_mp`.

woodsp-ibm commented:
I created an Issue in Finance - and linked it as you can see here above - since the docplex based applications there should be updated to make them compatible with this change.

t-imamichi commented:
I'm not sure why sphinx-spelling reports "October" is misspelling. But, I add it to pylintdict to pass CI, anyways.
https://github.com/Qiskit/qiskit-optimization/runs/2814027705
> /home/runner/work/qiskit-optimization/qiskit-optimization/docs/release_notes.rst:64: Spell check: october: which were deprecated in Qiskit Aqua 0.8.0 release (October 2020)..

manoelmarques commented:
As for having some template for deprecation msgs that @woodsp-ibm mentioned. I had created a set of methods in Qiskit-Aqua at https://github.com/manoelmarques/qiskit-aqua/blob/main/qiskit/aqua/deprecation.py.

I will create something similar to be used by the application projects.

t-imamichi commented:
Thank you. I fixed the deprecation message following @woodsp-ibm's advice. If your template is merged before the review of this PR, I use it.

manoelmarques commented:
@t-imamichi If you intend to finish this PR before Monday, feel free to do it. I will be looking at this template on Monday and looking at all repo usages so that it applies to everything. It may not be ready before you finish this PR.

t-imamichi commented:
I have finished the revision of PR just now. I need to wait for review.

woodsp-ibm commented:
Looks good but one final thing. If you look in the combined deprecation output in CI there are 4 warnings being emitted related to to/from docplex. Maybe these are from unit tests but I can't tell. I think normally we would have `stacklevel=2` on the warning so it refers to the caller - so it would be good to change the warnings to set that as it helps the end user too, who may have code they need to change.

If these warnings are from unit tests, normally there in the test we suppress and re-enable the warning around the code that we know will emit it so we don't end up seeing these anymore in CI. This is as ideally we should not be emitting deprecation warnings, by using Qiskit code, whether from other modules or tutorials, when we release. So avoiding warnings appearing from unit tests helps keep the log clean and easier to deal with the main code - now there are some warnings from 3rd party libraries, these are usually unavoidable until the library itself updates.

t-imamichi commented:
OK. The deprecation message was due to the unit test `test_docplex` of `from_docplex` and `to_docplex` themselves. I removed them.

woodsp-ibm commented:
Ok, normally we continue to test deprecated function while it exists to make sure it continues to work while we support it, ahead of its removal, but then suppress the deprecation warning around the test code as we know calling it will emit a warning. Anyway in this case I guess the risk is small in having no test coverage since the methods just simply call across to code that is tested.

It would be good even so though on the warnings.warn calls to have stacklevel=2 set so the site of where its called from is shown in the messages, otherwise it shows the function itself which can make it hard to find/know where its being called from.

t-imamichi commented:
I see. It is safer to keep `test_docplex`. I revert the removal and add stacklevel=2 to warnings.warn.

woodsp-ibm reviewed (approved):
I think we are good to go now. @t-imamichi thanks!
